Strategic model - general tutorial and treatment module Jupyter notebooks

tarnold17 opened the pull request:
## Fixes/Addresses/Summary/Motivation:

Contribute two Jupyter notebooks to the repo:
1) A general demonstration of the strategic model
2) A demonstration of the treatment module within the strategic model

In a subsequent PR, I plan to add some notes about these notebooks to the PARETO documentation.

## Changes proposed in this PR:
In addition to adding the Jupyter notebook, I have changed a few other small things in this PR:
- Refactor `generate_report` function so that the `is_print` argument does not have to be a list
- Update `generate_report` function so that the creation of an Excel file is optional (the Excel report is still generated by default, but now the user can pass `None` for the `fname` variable to prevent creating an Excel file)
- Remove the `has_pipeline_constraints` option from the config block for the strategic model (because it is not used anywhere in the model)
- Refactor the strategic model's `solve_model` function to improve handling of options
- Use the `PrintValues` enum in `generate_report` instead of hardcoding the values
- Fix a bug in the `handle_time` function which in certain situations causes `plot_sankey` to exclude some nodes from the generated Sankey diagram
- Allow the user to change Gurobi's `NumericFocus` parameter (issue #218)
- Fixing various typos and misspellings

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my
contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.md file
   at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has
   rights to intellectual property that includes these contributions, I represent that I have
   received permission to make contributions and grant the required license on behalf of that
   employer.

melody-shellman reviewed (commented):
Overall I think this will be useful for those interested in using the code! Perfect for onboarding new team members too! I have comments on mostly formatting things I noticed. 

1. Introduction - Not sure why, but the Strategic toy case study does not display properly on Chrome or Edge for me:
![image](https://user-images.githubusercontent.com/91741082/223844352-b526a414-486c-4a21-a44b-9d056ae38db8.png)

2. Introduction - Is it worth adding a note at the beginning that you must activate your PARETO conda environment for code run correctly? (Full disclosure, I forgot to haha)

3. Introduction - I think it would be nice to include a summary of the steps at the top with the introduction, so users know what to expect (1.Import files and libraries, 2. Set up....Step 8 (advanced) ...). 

4. Step 3 - Is it true that the values in the set_list and parameter_list variables are sometimes Pandas Dataframes? My understanding is that Dataframes are used for reading in and cleaning, but parameters are translated to dictionaries via the "to_dict()" function. I think it is that Sets are a Pandas Series because they are 1 dimensional and parameters are dictionaries. 

5. Step 4 - I think the formatting doesn't render correctly for me (tested on Chrome and Edge. 
![image](https://user-images.githubusercontent.com/91741082/223860934-e7c84988-7311-4d05-9eb0-fedb6216114c.png)

6. Step 4 - there is a typo in the code under "solver". Solver is spelled incorrectly in the comment. 

7. Step 6 - unscaled model units does not show with underscore?
![image](https://user-images.githubusercontent.com/91741082/223862310-57aad01e-0836-4c33-a699-44e1447912a1.png)

Sankey diagram section is great! It's a good way to break down build features we have and also help users get ideas for how to explore on the UI side. I do also really like the simple examples to change parameters and add constraints. 


solve_model changes looks good and is a good improvement to make code more readable!

tarnold17 commented:
I discussed with @MAZamarripa - he suggested to spin off the section on sensitivity analysis into a separate Jupyter notebook which is more focused on the treatment module within the strategic model.

lbianchi-lbl commented:
This is being reviewed offline and will have to be included in the March release.

tarnold17 commented:
@melody-shellman - thanks for your great suggestions. I believe I have addressed all of your comments with my latest commits.

Regarding this comment:

> Step 3 - Is it true that the values in the set_list and parameter_list variables are sometimes Pandas Dataframes?

The wording at the beginning of section 3 is admittedly a bit cryptic, but I am not claiming that the values in `set_list` and `parameter_list` are sometimes dataframes. Rather, I am saying that the values in `df_sets` and `df_parameters` are sometimes dataframes.

The missing underscores were not actually missing, just being covered up by other formatting. One thing I have learned from this PR is that there is not an easy way to include lists within tables in Markdown. My solution is a bit of a hack, but the formatting should look better now.

Note that sections 7 and 8 from the Jupyter notebook have been removed - I have moved them to a separate Jupyter notebook for which I am still finishing up development.

melody-shellman commented on pareto/examples/strategic_model_treatment_module_demo/strategic_model_treatment_module_demo.ipynb:
- [ ] Treatment Technologies: May be good to note that users can include additional technologies. The name can be any that user chooses, but references to technology in other tabs must match the name.
- [ ] Treatment expansion cost: The Note at bottom is formatted differently. Other notes start with "NOTE:"
- [ ] Component removal efficiency: this will need update with PR #228 changes. It may be good to include a brief description of config option and link to documentation in PR #229 .
- [ ] I was unable to run the first cell in the jupyter notebook because of the line "from tabulate import tabulate". I know we talked about this briefly, but I may have missed it, should we update the requirements with this library? I also suggest adding the import for the tabulate library to the first cell that uses it. While it's nice to include all imports together, I feel there's less room for error and the library function is more clear if it's in the relevant cell. I was able to run it once I commented it out. 
- [x]  The plots are interesting and effective way to explore varying OPEX for MVC
- [x] The interactive tool looks awesome! I didn't know you could do this in jupyter notebooks. A couple of comments: 
![image](https://user-images.githubusercontent.com/91741082/231836701-80b75b94-f8ac-436b-b7bb-f06908166a82.png)
- [ ] The formatting of "choose desalination for R01" doesn't display correctly, so is a bit confusing. Lastly, to me it seems user will not be interested in the set-up for the interactive section. Do you think it might be worth it to create a python script for jupyter notebooks that can pull in the methods you created? Note: I didn't run this tool because I don't have the tabulate library. I also don't want to install it unless it is through the requirements file, so I can test that the version we include works.

tarnold17 replied:
More good suggestions here. I will respond only to the most significant comments from above, but I plan to address all of the suggestions.

You are right that we should add the "tabulate" package to the requirements file since I am using it in this notebook. @lbianchi-lbl is that something we can do as part of this PR? I just ran `pip install tabulate` in my PARETO conda environment to get it. @melody-shellman I think you could do the same without any negative repercussions if you want to continue testing the notebook.

It's curious to me that the formatting of the widgets is incorrect for you. Here's what I see on my machine:

![image](https://user-images.githubusercontent.com/26798258/232582532-b333f0e5-3a9a-4ea1-bf72-6fdbf1c9379f.png)

We should investigate why it's rendering differently for you.

> Lastly, to me it seems user will not be interested in the set-up for the interactive section. Do you think it might be worth it to create a python script for jupyter notebooks that can pull in the methods you created?

I think this is a good point and a good idea. I will look into spinning off all of the widget generation code into a separate file so that the notebook user doesn't have to use it if they don't want to.

tarnold17 replied:
I believe I have addressed all the points here, so I am resolving.

melody-shellman commented on pareto/utilities/results.py:
Good general improvements here.

melody-shellman reviewed (commented):
Great improvements! I've included a few minor comments on formatting, some small suggested additional notes.  The larger suggestions that I have (also included in specific comments throughout) are:

- First, I suggest that changes from PR #228 and PR #229 should inform a few updates to this notebook. 
- Second, there are a number of output files (excel, html) that I believe can be removed from the repo.
- Lastly, the library tabulate needs to be added to the requirements. I think it should be added within this PR.

tarnold17 commented on pareto/examples/strategic_model_general_demo/Section1_demo_sankey_T01_sections.html:
> It looks like these are output from the run file
These files are output from running the notebook, yes.

> I suspected it may be so that users don't have to to run the notebook
This is precisely my intention for including all of the output files in the repo alongside the Jupyter notebooks themselves. This is a question we might want to ask of the rest of the team, however - when adding Jupyter notebooks to the repo, should the `ipynb` file be at its "end" state (i.e., after it has been run) along with all outputs? Or should just the base, "vanilla" notebook be included? I would be curious what @ksbeattie, @dangunter, and @lbianchi-lbl have to say on that question, as I know they have worked on other projects (namely, IDAES) where Jupyter notebooks have been developed and then added to a repo of examples.

> it looks to me that having the image in the folder doesn't mean that it will show in the notebook. Is this incorrect?
By "image", you are still referring to the image contained within the `*.html` files, correct? If I am understanding your question properly, then I think you are incorrect - e.g., when I run `jupyter notebook strategic_model_general_demo.ipynb`, all of the plots do appear in the notebook once it loads.

tarnold17 replied:
After discussing with @lbianchi-lbl and @ksbeattie, we have decided (for now) to merge the unexecuted version of Jupyter notebooks into the repo. All of the output files have been removed from this PR now, so I am resolving this comment.

tarnold17 commented on pareto/examples/strategic_model_general_demo/strategic_toy_network.png:
It's possible, yes, but not necessarily desirable, in my opinion. My intention of duplicating the file was so that each folder with a Jupyter notebook can be self contained. That is, this way, we could zip up just this folder and distribute it separately from the rest of the repo, and a user would be able to take that and run it (as long as they have PARETO installed).

tarnold17 commented on pareto/examples/strategic_model_treatment_module_demo/strategic_toy_network.png:
See above response.

tarnold17 commented on pareto/examples/strategic_model_treatment_module_demo/data_input_treatment_removal_efficiency.png:
We should update the image, you are right. However, I think I will wait to do so until #228 is merged (which should be very soon).

tarnold17 replied:
Addressed.

tarnold17 commented on pareto/examples/strategic_model_general_demo/strategic_optimization_results.xlsx:
It's definitely not necessary to include, but if we want the folder containing the Jupyter notebook to essentially be a "snapshot" of the state after running the notebook, then I think it makes sense to include this file.

tarnold17 replied:
This file has been removed from the PR, so I am resolving.

tarnold17 commented on pareto/examples/strategic_model_general_demo/strategic_model_general_demo.ipynb:
Addressed.

melody-shellman commented on pareto/examples/strategic_model_treatment_module_demo/utilities.py:
This is great! A general note - as we add more examples/jupyter notebooks, they may also utilize some of these utilities. For now, I think we can keep as is, but as more are added, this file may benefit from being located in the examples directory.

tarnold17 replied:
These utilities were created specifically for the treatment module Jupyter notebook, so I'm not sure that there is anything in there which would be relevant or useful anywhere else. But I do agree with the comment that we may want to develop some centralized utilities to be used in Jupyter notebooks as we add more of them to the repo.

I'm resolving the comment since there is no action item.

melody-shellman commented on pareto/examples/strategic_model_general_demo/strategic_model_general_demo.ipynb:
Really minor comment- Step 4: code block, for consistency can we also add the removal_efficiency_method to the create_model() call?
Otherwise, looking really good!

tarnold17 replied:
Addressed!

melody-shellman commented on pareto/examples/strategic_model_treatment_module_demo/strategic_model_treatment_module_demo.ipynb:
Another minor comment: Sensitivity Analysis plot summary: The units of v.Z is in k$ as opposed to $ ! Just an update the the y_label needed. It may also be worth noting in the model interaction section or update the create_widgets() code to display k$ as units for objective function value. 
Love the interactive feature! so cool!

melody-shellman reviewed (commented):
Looking good! A few minor comments noted in the .ipynb files. 
Lastly, a question - Did tabulate library get removed ? I didn't seem to run into any issues with running cells this go around.

tarnold17 commented:
> Looking good! A few minor comments noted in the .ipynb files. Lastly, a question - Did tabulate library get removed ? I didn't seem to run into any issues with running cells this go around.

Yes, the tabulate library did get removed. Ludovico helped me find a way to accomplish the same thing using pandas instead.

melody-shellman reviewed (dismissed):
Looks good!

nareshsusarla commented on pareto/examples/strategic_model_treatment_module_demo/strategic_model_treatment_module_demo.ipynb:
```diff
@@ -0,0 +1,1662 @@
+   "metadata": {},
+   "source": [
+    "### Component removal effiency\n",
+    "In tab \"RemovalEfficiency\", for each, site, technology, and quality component, specify the fraction of the component (by mass) which goes to the residual stream:\n",
```
A recall comment to differentiate removal efficiency from the treatment efficiency defined above may be useful here

tarnold17 replied:
Addressed.

nareshsusarla commented on pareto/examples/strategic_model_treatment_module_demo/strategic_model_treatment_module_demo.ipynb:
```diff
@@ -0,0 +1,1662 @@
+    "## <a id='solve'></a>Set up and solve Pyomo model\n",
+    "To begin, ensure that you are working in a development environment where PARETO is installed. For example, if you followed the PARETO [installation instructions](https://pareto.readthedocs.io/en/latest/getting_started/index.html) to create a conda environment called `pareto-env`, then make sure that environment is active by running `conda activate pareto-env`.\n",
+    "\n",
+    "NOTE: The code in this notebook uses the Gurobi commerical optimization solver by default. If you do not have Gurobi installed, it will use the free solver CBC instead. The code will be much slower running with CBC than with Gurobi.\n",
```
somehow i feel we should remove `much` from the last line for solver classification

tarnold17 replied:
Addressed.

nareshsusarla commented on pareto/examples/strategic_model_treatment_module_demo/strategic_model_treatment_module_demo.ipynb:
```diff
@@ -0,0 +1,1662 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "At site R01, the optimizer has chosen to build MVC desalination with the capacity selected corresponding to the `J1` index. As an aside, we can look at the input data to remind ourselves what capacity `J1` corresponds to (in bbl/day):"
```
Again, the flow of instructions assumes the cells are run. So, in my opinion, it may be useful to save the notebook after cells are run. The read through will be more meaningful with references being apparent.

tarnold17 replied:
See my above comment on the matter. If we want to incorporate the notebook into the documentation, it will need to be an executed version of the notebook. But for now, we are just going to commit uncommitted notebooks to the repo until we figure out exactly how we want to incorporate the notebooks into the documentation.

tarnold17 commented on pareto/examples/strategic_model_general_demo/strategic_model_general_demo.ipynb:
```diff
@@ -0,0 +1,540 @@
+{
```
We discussed this with the Berkeley guys and decided for now to commit unexecuted versions of the Jupyter notebook files to the repo. There are pros and cons to doing it either way. If we decide later that we want executed versions of the notebooks in the repo, we can update them.

tarnold17 commented on pareto/examples/strategic_model_general_demo/strategic_model_general_demo.ipynb:
```diff
@@ -0,0 +1,540 @@
+   "id": "01914afa",
+   "metadata": {},
+   "source": [
+    "PARETO's `is_feasible` utility function can be called after solving the model to check whether or not the solution returned by the solver violates any constraints:"
```
Can you give a more specific recommendation? I feel like the text already address when to call `is_feasible` ("after solving the model") and why it's useful ("to check whether or not the solution returned by the solver violates any constraints").

nareshsusarla replied:
I understood that some solvers may not return status correctly if they didn't solve to optimality. So I agree with keeping this check for now for all solution as an additional check. Resolving this comment in this regard.

nareshsusarla reviewed (approved):
LGTM

melody-shellman reviewed (approved):
Looks good!
